src/ast.rs

```diff
@@ -2457,7 +1921,7 @@ pub struct Extras {
 pub struct Ast {
     // The top node.
     // Its type depends on what was requested to parse.
-    top: Box<dyn NodeMut>,
+    top: Box<NodeEnum>,
```

Comment:
Does this still need to be boxed?
I guess we still need a stable address for `parent` but we should probably be using `Pin` for that.
In doubt we can leave it until we pin

Reply:
Good eye. I'm pretty sure top level will still need a `Box`, so that moving an `Ast` doesn't invalidate the first level descendants. None of the other boxes are necessary, but I'm trying to limit the scope of each of these PRs.

I believe a fully idiomatic implementation would use `Pin<Box<T>>` for the root and `Option<Pin<NonNull<T>>>` for the children, as well as a `PhantomPinned` marker field in every node that has children.

However, `Pin` is mostly a guardrail. I think we already establish the necessary guarantees by:
- not exposing `&mut` access once parsing is complete
- not populating parent pointers until the tree is constructed

I plan on looking into `Pin`, but it may or may not prove cumbersome to work with here.